opentelemetry-api/src/opentelemetry/context/async_context.py

```diff
@@ -17,29 +17,54 @@
 except ImportError:
     pass
 else:
-    import typing  # pylint: disable=unused-import
+    # import contextvars
+    import typing
     from . import base_context
 
-    class AsyncRuntimeContext(base_context.BaseRuntimeContext):
-        class Slot(base_context.BaseRuntimeContext.Slot):
-            def __init__(self, name: str, default: object):
-                # pylint: disable=super-init-not-called
-                self.name = name
-                self.contextvar = ContextVar(name)  # type: ContextVar[object]
-                self.default = base_context.wrap_callable(
-                    default
-                )  # type: typing.Callable[..., object]
-
-            def clear(self) -> None:
-                self.contextvar.set(self.default())
-
-            def get(self) -> object:
-                try:
-                    return self.contextvar.get()
-                except LookupError:
-                    value = self.default()
-                    self.set(value)
-                    return value
-
-            def set(self, value: object) -> None:
-                self.contextvar.set(value)
+    class ContextVarSlot(base_context.Slot):
+        def __init__(self, name: str, default: object):
+            # pylint: disable=super-init-not-called
+            self.name = name
+            self.contextvar = ContextVar(name)  # type: ContextVar[object]
+            self.default = base_context.wrap_callable(
+                default
+            )  # type: typing.Callable[..., object]
+
+        def clear(self) -> None:
+            self.contextvar.set(self.default())
+
+        def get(self) -> object:
+            try:
+                return self.contextvar.get()
+            except LookupError:
+                value = self.default()
+                self.set(value)
+                return value
+
+        def set(self, value: object) -> None:
+            self.contextvar.set(value)
+
+    class AsyncRuntimeContext(base_context.Context):
+        def with_current_context(
```

Comment:
I'm wondering if this method has to be different for both implementations.

Wouldn't an implementation like the following work? 
https://github.com/open-telemetry/opentelemetry-python/blob/ccb97e582555391600ac45c44be2eb61b522b61f/opentelemetry-api/src/opentelemetry/context/base_context.py#L112-L130